text/3366-diagnostic-attribute-namespace.md

```diff
@@ -0,0 +1,222 @@
+
+The second user group interacts with the described feature through attributes. These attributes allow them to hint the compiler to emit specific error messages in certain cases. The `#[diagnostic]` attribute namespace provides a general framework for what can and can't be done by such an attribute. As of this users won't interact directly with the attribute namespace itself. 
+
+The `#[diagnostic::on_unimplemented]` attribute allows to hint the compiler to emit a specific error message if a certain trait is not implemented. This attribute should provide the following interface:
+
+```rust
+#[diagnostic::on_unimplemented(
+    message="message",
+    label="label",
+    note="note"
+)]
+trait MyIterator<A> {
+    fn next(&mut self) -> A;
+}
+
+
+fn iterate_chars<I: MyIterator<char>>(i: I) {
+    // ...
+}
+fn main() {
+    iterate_chars(&[1, 2, 3][..]);
+}
+```
+
+which might result in the compiler emitting the following error message:
+
+```
+error[E0277]: message
+  --> <anon>:14:5
+   |
+14 |     iterate_chars(&[1, 2, 3][..]);
+   |     ^^^^^^^^^^^^^ label
+   |
+   = note: note
+   = help: the trait `MyIterator<char>` is not implemented for `&[{integer}]`
+   = note: required by `iterate_chars`
+```
+
+I expect the new attributes to be documented on the existing [Diagnostics](https://doc.rust-lang.org/reference/attributes/diagnostics.html) attributes page on the rust reference similar to existing attributes like for example `#[deprecated]`
+
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+## The `#[diagnostic]` attribute namespace
+
+This RFC proposes to introduce a new `#[diagnostic]` attribute namespace. This namespace is supposed to contain different attributes, which allow users to hint the compiler to emit specific error messages in certain cases like type mismatches, unsatisfied trait bounds or similar situations. By collecting such attributes in a common namespace it is easier for users to find useful attributes and it is easier for the language team to establish a set of common rules for these attributes.
+
+Any attribute in this namespace may:
+
+* Hint the compiler to emit a specific error message in a specific situation
+
+Any attribute in this namespace is not allowed to:
+
+* Change the result of the compilation, which means applying such an attribute should never cause an error as long as they are syntactically valid
+
+The compiler is allowed to:
+
+* Ignore the provided hints
+* Use only parts of the provided hints, for example for the proposed `#[diagnostic::on_unimplemented]` only use the `message` option
+* Change this behaviour at any time
```

Comment:
I am happy with api evolution, but I'd like to be aware if/when this happens. I don't expect it to actually change much (after all, `rustc_on_unimplemented` has barely changed backwards compatibly), but I am a bit afraid of having a lot of these attributes in my codebase, only for future me to have that "oh this hasn't actually worked for a long time" moment.

Perhaps a lint diagnostic should be emitted for that?

Reply:
`rustc_on_unimplemented` hasn't ever broken backwards compatibility, somehow 😅

 > Perhaps a lint diagnostic should be emitted for that?

+1

Reply:
I personally would consider that a quality of implementation issue. Which means that's something where a specific compiler implementation can offer a lint for. Other than that: At least for my own use-cases I will likely have tests based on `trybuild` or something else. These will show whether the attributes are ignored or not.